cpp/include/rapidsmpf/streaming/cudf/shuffler.hpp

```diff
@@ -5,36 +5,202 @@
 
 #pragma once
 
+#include <unordered_set>
+
 #include <rapidsmpf/shuffler/shuffler.hpp>
 #include <rapidsmpf/streaming/core/channel.hpp>
 #include <rapidsmpf/streaming/core/context.hpp>
 #include <rapidsmpf/streaming/core/node.hpp>
 #include <rapidsmpf/streaming/cudf/partition.hpp>
 
-namespace rapidsmpf::streaming::node {
+namespace rapidsmpf::streaming {
+
+/**
+ * @brief An asynchronous shuffler that allows concurrent insertion and extraction of
+ * data.
+ *
+ * ShufflerAsync provides an asynchronous interface to the shuffler, allowing data to be
+ * inserted while previously shuffled partitions are extracted concurrently. This is
+ * useful for streaming scenarios where data can be processed as soon as individual
+ * partitions are ready, rather than waiting for the entire shuffle to complete.
+ *
+ * The class maintains internal synchronization using coroutine-based mutexes and
+ * condition variables to coordinate between insert and extract operations safely.
+ *
+ * @note This class is non-copyable to prevent accidental duplication of the underlying
+ * shuffler state and synchronization primitives.
+ */
+class ShufflerAsync {
+  public:
+    /**
+     * @brief Constructs a new ShufflerAsync instance.
+     *
+     * @param ctx The streaming context containing communicator, progress thread, and
+     * other shared resources.
+     * @param stream The CUDA stream on which shuffling operations will be performed.
+     * @param op_id Unique operation ID for this shuffle. Must not be reused until all
+     *              participants have completed the shuffle operation.
+     * @param total_num_partitions Total number of partitions to shuffle data into.
+     * @param partition_owner Function that maps a partition ID to its owning rank/node.
+     *                        Defaults to round-robin distribution.
+     */
+    ShufflerAsync(
+        std::shared_ptr<Context> ctx,
+        rmm::cuda_stream_view stream,
+        OpID op_id,
+        shuffler::PartID total_num_partitions,
+        shuffler::Shuffler::PartitionOwner partition_owner =
+            shuffler::Shuffler::round_robin
+    );
+
+    // Prevent copying
+    ShufflerAsync(const ShufflerAsync&) = delete;
+    ShufflerAsync& operator=(const ShufflerAsync&) = delete;
+
+    ~ShufflerAsync() = default;
+
+    /**
+     * @brief Gets the streaming context associated with this shuffler.
+     *
+     * @return A reference to the shared context object.
+     */
+    constexpr auto& ctx() const {
```

Comment:
No `auto` in the public API
```suggestion
    constexpr std::shared_ptr<Context>& ctx() const {
```